Mixed effects model to convert irregular data to basis expansion

pcuestas opened the pull request:
## References to issues or other PRs

* This branch includes the changes not yet merged into develop from https://github.com/GAA-UAM/scikit-fda/pull/610

* Problems found when evaluating FDataIrregular objects in some cases are described in #616 and #617 

## Describe the proposed changes


### 1. Mixed-effects model

Implementation of the mixed effects method for converting irregularly sampled functional data to basis expansion.

This implementation includes the addition of a new module: skfda.representation.conversion, which is meant for including classes (transformers) that convert between different FData sublasses.

For this type of conversion (FDataIrregular to FDataBasis) two classes have been added: EMMixedEffectsConverter and MinimizeMixedEffectsConverter, each of which implements the conversion with a different method to fit the MLE of the mixed-effects model.

The EMMixedEffectsConverter uses the EM algorithm and the MinimizeMixedEffectsConverter uses generic optimizers from scipy.optimize.minimize to find maxima of the loglikelihood function.


### 2. Create an FDataIrregular object from another FData

A function: `irregular_sample` has been added to skfda.datasets, which creates an irregular sample by randomly selecting points from another FData object.

## Additional information

All implementation added has been adapted (and tested) for multidimensional domains and codomains.

However, some problems have been found regarding this topic and are described in #616 and #617 .

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] The code conforms to the style used in this package
- [x] The code is fully documented and typed (type-checked with [Mypy](https://mypy-lang.org/))
- [x] I have added thorough tests for the new/changed functionality

vnmabus commented on skfda/tests/test_mixed_effects_converter.py:
```diff
@@ -0,0 +1,281 @@
+    basis = FourierBasis(n_basis=5, domain_range=(0, 10))
+    model = _MixedEffectsModel(fdatairregular, basis)
+
+    # These values have been obtained with Statsmodels' MixedLM
```
It would have been good to have in the docstring the lines of code that generate these data, if possible.

vnmabus reviewed (changes requested):
Hi @pcuestas, I hope everything is ok. First, I have to apologize for taking so much time in reviewing your work in depth. In general it is *really* good, but I have found some minor things that could be improved.

Of course, I would totally understand if you do not have the time or will to review this after so much time. In that case, please just tell me and I will try to do the changes myself and merge it.

pcuestas commented:
Hello @vnmabus, I also hope you are doing great. Thanks for the review. I am very occupied with the master's until March 29th. I can do the changes then if waiting for a couple of weeks is OK.

vnmabus commented:
I am glad that you are doing well! Of course it is ok, take the time you need. In case you change your mind, again just tell me.

vnmabus commented:
Hey, @pcuestas. I hope things are better now with your master. Do you think it would be possible to have this merged in the next days/weeks? If you cannot do the work it is fine, I can finish it, but please tell me (and answer the questions if possible).

I am sorry about asking, but I want to do a release in the next few days and I would prefer to have this PR merged for that.

pcuestas commented on skfda/representation/conversion/_mixed_effects.py:
```diff
@@ -0,0 +1,1046 @@
+    )  # type: ignore
+
+
+class _MixedEffectsParams(Protocol):
```
There is more than one implementer, e.g. _MixedEffectsParamsResult, MinimizeMixedEffectsConverter.Params and EMMixedEffectsConverter.Params.

vnmabus replied:
Ok, I didn't know. I would prefer them to inherit explicitly from the protocol, so that it is more clear, but it seems that it is [currently not possible](https://github.com/python/cpython/issues/91393). Until that changes I will leave that as is.

pcuestas commented:
Hi again, @vnmabus. Sorry for the delay. I changed setup since I wrote the code and between that and the changes undergone in the package, now that I started to run the tests again I am having trouble with dependencies versions. For this reason, I think it is best if you finish the remaining tasks that you have commented on, otherwise I'll spend too much time trying to fix dependencies. 

I have done some of the suggested changes which do not significantly change the code (mainly comments, because I can't fully run the tests). In particular, I added the code that I used to compute the loglikelihood from statsmodels' MixedLM. I hope this is fine and I am sorry that I cannot do much more than that. Please reach out if you have any question.

vnmabus commented:
Thank you so much! Actually what you did are probably the most important changes for me, as are the ones that I could not do on my own.

It is possible that the dependencies are fixed in the `develop` branch. In any case, I will try to finish it, and if I have any question I will ask you.

vnmabus reviewed (approved):
Ok, I did the finishing touches. Thank you @pcuestas for your hard work!

vnmabus commented:
@all-contributors Please add @pcuestas for code, documentation, examples, tests, and research.
